pgscm/farmer/views.py

```diff
@@ -22,35 +26,84 @@ def index():
+                                         models.Group.name.asc()).all()]
+
+        # form create or edit submit
+        if request.method == 'POST' and form.data['submit']:
+            if not check_role():
+                return redirect(url_for(request.endpoint))
+            elif form.validate_on_submit():
+                # edit user
+                if form.id.data:
+                    edit_farmer = sqla.session.query(models.Farmer) \
+                        .filter_by(id=form.id.data).one()
+                    edit_farmer.farmer_code = form.farmer_code.data
+                    edit_farmer.name = form.name.data
+                    edit_farmer.gender = form.gender.data
+                    edit_farmer.type = form.type.data
+                    if edit_farmer.group_id != form.group_id.data:
+                        new_group = models.Group.query.filter_by(
+                            id=form.group_id.data).one()
+                        edit_farmer.group = new_group
+                    sqla.session.commit()
+                    for fm in farmers:
+                        if fm.id == edit_farmer.id:
+                            farmers.remove(fm)
+                            farmers.append(edit_farmer)
+                    flash('Update farmer success!', 'info')
+
+                # add user
+                else:
+                    group_farmer = models.Group.query.filter_by(
+                        id=form.group_id.data).one()
+                    new_farmer = models.Farmer(type=form.type.data,
+                        farmer_code=form.farmer_code.data, name=form.name.data,
+                        group=group_farmer, gender=form.gender.data)
+                    sqla.session.add(new_farmer)
+                    sqla.session.commit()
+                    farmers.append(new_farmer)
+                    flash('Add farmer success!', 'info')
+
+                return redirect(url_for(request.endpoint))
+
+        # form delete submit
+        if request.method == 'POST' and dform.data['submit_del']:
+            if not check_role():
+                return redirect(url_for(request.endpoint))
+            elif dform.validate_on_submit():
+                del_farmer = sqla.session.query(models.Farmer) \
+                    .filter_by(id=dform.id.data).one()
+                del_farmer._deleted_at = func.now()
+                if dform.modify_info.data:
+                    del_farmer._modify_info = dform.modify_info.data
+                sqla.session.commit()
+                farmers.remove(del_farmer)
+                flash('Delete farmer success!', 'info')
+
+                return redirect(url_for(request.endpoint))
 
         return render_template('farmer/index.html', farmers=farmers,
                                form=form, dform=dform)
+
+
+def check_role():
+    for r in crud_role:
+        if r == current_user.roles[0].name:
+            return True
+    flash('You have no permission!', 'warning')
```

Comment:
i18n, the correct is `flash('warning', 'no permission')` and delete block of code in template regarding flash message.